File: Days.31/C++/2.Reverse-DLL-in-a-group-of-k.cpp
// Reverse Doubly Linked List in a group of size k

#include<iostream>
using namespace std;

class Node
{
    public:
    int val;
    Node *prev;
    Node *next;
    Node(int val)
    {
        this->val=val;
        this->prev=this->next=nullptr;
    }
};

class DLL
{
    Node *head,*tail;
    public:
    DLL()
    {
        head=tail=nullptr;
    }
    void Insert(int val)
    {
        Node *temp=new Node(val);
        if (head==nullptr)
        {
            head=tail=temp;
        }
        else
        {
            tail->next=temp;
            temp->prev=tail;
            tail=temp;
        }
    }

    void Reverse(int k)
    {
        Node *oh,*ot,*th,*tt,*curr,*next;
        oh=ot=th=tt=next=nullptr;
        curr=head;
        int count=0;
        while (curr!=nullptr)
        {
            next=curr->next;
            if(count==k)
            {
                if (oh==nullptr)
                {
                    oh=th;
                    ot=tt;
                }
                else
                {
                    ot->next=th;
                    th->prev=ot;
                    ot=tt;
                }
                count=0;
                th=tt=nullptr;
            }
            else
            {
                if(th==nullptr)
                {
                    curr->prev=curr->next=nullptr;
                    th=tt=curr;
                }
                else
                {
                    curr->next=th;
                    th->prev=curr;
                    curr->prev=nullptr;
                    th=curr;
                }
                count+=1;
                curr=next;
            }
        }
        if(th!=nullptr)
        {
            ot->next=th;
            th->prev=ot;
            ot=tt;
        }
        head=oh;
        tail=ot;
    }

    void printHead()
    {
        Node *curr=head;
        while (curr!=nullptr)
        {
            cout<<curr->val<<" ";
            curr=curr->next;
        }
        cout<<"\n";
    }
    void printTail()
    {
        Node *curr=tail;
        while (curr!=nullptr)
        {
            cout<<curr->val<<" ";
            curr=curr->prev;
        }
        cout<<"\n";
    }
};

void BuildDLL(DLL *&dll,string &str)
{

    string temp="";
    for(int i=0; i<str.size(); i++)
    {
        if(str[i]==' ')
        {
            dll->Insert(stoi(temp));
            temp="";
        }
        else{
            temp+=str[i];
        }
    }
    dll->Insert(stoi(temp));
}

int main()
{
    string str;
    getline(cin,str);
    int k;
    cin>>k;
    DLL *dll=new DLL();
    BuildDLL(dll,str);
    dll->printHead();
    dll->printTail();
    dll->Reverse(k);
    dll->printHead();
    dll->printTail();
    return 0;
}

File: Days.31/Python/2.Reverse-DLL-in-a-group-of-k.py
# Reverse a Doubly Linked List in a group of given size k

class Node:
    def __init__(self,val):
        self.val=val
        self.prev=None
        self.next=None

class DLL:
    def __init__(self):
        self.head=self.tail=None
    def Insert(self,val):
        if self.head==None:
            self.head=self.tail=Node(val)
        else:
            temp=Node(val)
            self.tail.next=temp
            temp.prev=self.tail
            self.tail=temp 
    '''
    ///Concept: Reverse a Linked List using AddFirst 
        struct node *th,*tt,*oh,*ot;
        oh=ot=th=th=NULL;
        int count;
        struct node *c,*f;
        c=head;
        count=k;
        while(c!=NULL)
        {
            f=c->next;
            if(th==NULL)
            {
                th=tt=c;
                c->next=NULL;
            }
            else
            {
                c->next=th;
                th=c;
            }
                count--;
            if(count==0)
            {
                if(oh==NULL)
                {
                    oh=th;
                    ot=tt;
                    th=tt=NULL;
                }
                else
                {
                    ot->next=th;
                    ot=tt;
                    th=tt=NULL;
                }
                count=k;
            }
            c=f;
        }
        if(th!=NULL)
        {
            ot->next=th;
            ot=tt;
        }
        return oh;
    '''
    def Reverse(self, k):
        oh=ot=None
        th=tt=None
        count=0
        curr=self.head
        next=None
        while curr!=None:
            next=curr.next
            if count==k:
                if oh==None:
                    oh=th
                    ot=tt
                else:
                    ot.next=th
                    th.prev=ot
                    ot=tt
                th=tt=None
                count=0
            else:
                if th==None:
                    curr.prev=curr.next=None
                    th=tt=curr
                else:
                    curr.next=th
                    th.prev=curr
                    curr.prev=None
                    th=curr
                count+=1
                curr=next
        if th!=None:
            ot.next=th
            th.prev=ot
            ot=tt
        self.head=oh
        self.tail=ot

    def printHead(self):
        curr=self.head
        while curr!=None:
            print(curr.val,end=" ")
            curr=curr.next
        print()
    def printTail(self):
        curr=self.tail
        while curr!=None:
            print(curr.val,end=" ")
            curr=curr.prev
        print()
    

def main():
    nodes=input().split()
    k=int(input())
    dll=DLL()
    for node in nodes:
        dll.Insert(int(node))
    dll.printHead()
    dll.printTail()
    dll.Reverse(k)
    dll.printHead()
    dll.printTail()


main()

